extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -392,39 +420,44 @@ \section{EUTXO-1: Enhanced scripting}
-  which makes it easier to have an ``identity'' for an ongoing
-  contract: we will return to this issue in
-  Section~\ref{sec:validating-datascripts}.
+\item Each unspent output now has a (possibly empty) object $\valdata$ of type
+  \Data{} associated with it: we (\red{provisionally}) call this the output's
+  \textit{validation data}.  The introduction of the validation data
+  increases the expressivity of the model considerably. For example,
+  one can use the validation data to propagate state between
+  transactions, and this can be used to give a contract the structure
+  of a finite state machine; the fact that the validation data is part
+  of the output and not the transaction means that the state can
+  change without the transaction changing, which makes it easier to
+  have an ``identity'' for an ongoing contract.
+
+\item The redeemer script of~\citet{Zahnentferner18-UTxO} has been
+  replaced with a redeemer object of type \Data.
   
-\item Validation is performed by running the validator script with the
-  redeemer and data script as input; the validator is also provided
-  with information about the pending transaction (ie, the transaction
-  which is just about to take place, assuming that validation
-  succeeds); this information is contained in a structure which we call
-  \ptx{} which contains data such as the validity interval of the
-  pending transaction and information about the inputs and outputs.
-  See Section~\ref{sec:pendingtx} for more information on the $\ptx$
-  structure.
+\item Validation of an output is performed by running the validator
+  script with the redeemer and validation data as input; the validator
+  is also provided with information about the pending transaction (ie,
+  the transaction which is just about to take place, assuming that
+  validation succeeds); this information is contained in a structure
+  which we call \ptx{} which contains data such as the validity
+  interval of the pending transaction and information about the inputs
+  and outputs.  See Section~\ref{sec:pendingtx} for more information
+  on the $\ptx$ structure.
   
 \end{itemize}
 
 
 \noindent At the ledger level, validator scripts now have the type
-$$
-\script \times \script \times \script \rightarrow \B
-$$
-At the Plutus level in Cardano, the validator is a program which
-consumes the redeemer, the data script, and the Plutus Core encoding of
-the $\ptx$ structure as arguments: each script will have a Plutus type
-and the types will have to fit the types expected by the validator.
+$
+\Data \times \Data \times \Data \rightarrow \B
+$. At the Plutus level in Cardano, the validator is a program which
+consumes the redeemer, the validation data, and a \Data{} encoding
+of the $\ptx$ structure as arguments.  The validator may
+expect these objects to have some specific form (a single bytestring
+or a list of integers, for example) and if it is supplied with data
+with an incorrect format it should fail and return \false.
```

Comment:
There's an awkwardness here about whether scripts return `Bool` or `()` with failure indicated by error. If it's `Bool` then we have to explain how the ledger layer interprets the output, which is tricky. What we do at the moment is the ledger applies a special function `check` that just throws on false.

However, I think really the ledger should see the interface where it returns `()` and can fail (or indeed, the return type is irrelevant and should be ignored). That way we keep the signalling to easily observed parts of evaluation.

To facilitate this I'm planning to move the application of `check` into the function that wraps validators.

Reply:
I'll merge the part about script return values as is for now, but I've added a reminder to come back and fix it.